### PR TITLE
Fix Jacoco to work with build cache

### LINKED ISSUE #5269: Support the build cache and parallel test execution with the JaCoCo plugin applied
### Update
The wider implications of this bug is captured well here:
https://github.com/gradle/gradle/issues/5269#issuecomment-390489401

### Expected Behavior
If you use the Jacoco plugin with it's default configuration, it should not break the `Test` task's ability to be cached.

### Current Behavior
Applying the `Jacoco` plugin to your project makes the `Test` task not support caching.

https://github.com/gradle/gradle/blob/76387d95bd2d785eb76991d252102b9546551696/subprojects/jacoco/src/main/java/org/gradle/testing/jacoco/plugins/JacocoPluginExtension.java#L131-L136

https://docs.gradle.org/current/dsl/org.gradle.testing.jacoco.plugins.JacocoTaskExtension.html#org.gradle.testing.jacoco.plugins.JacocoTaskExtension:append

> Whether or not data should be appended if the destinationFile already exists. Defaults to true.

### Context
It makes sense that if you have custom configuration for the `Jacoco` plugin you might prevent Build Caching from working, but in it's default configuration it should work.

### Steps to Reproduce (for bugs)
Add the `Jacoco` plugin to a project and be using the Build Cache.

### Workaround
Using the Kotlin DSL:
```kotlin
tasks.withType<Test> {
    extensions.configure(typeOf<JacocoTaskExtension>()) {
        isAppend = false
    }
}
```

### Your Environment
```
------------------------------------------------------------
Gradle 4.7
------------------------------------------------------------

Build time:   2018-04-18 09:09:12 UTC
Revision:     b9a962bf70638332300e7f810689cb2febbd4a6c

Groovy:       2.4.12
Ant:          Apache Ant(TM) version 1.9.9 compiled on February 2 2017
JVM:          1.8.0_92 (Oracle Corporation 25.92-b14)
OS:           Mac OS X 10.13.3 x86_64

```

### PR DESCRIPTION
**Pull request by wolfs**
Jacoco code coverage should work well with the build cache out of the
box. Since appending to a coverage file works with parallel test
execution, see https://github.com/jacoco/jacoco/pull/52, we set
`append=true` and delete the coverage data just before the test task
starts to execute.

Note that this is a breaking change: separate tasks now cannot use the
same coverage file, since each of the tasks will delete it.

Fixes #5269

### CODE REVIEW

File: subprojects/jacoco/src/main/java/org/gradle/testing/jacoco/plugins/JacocoPluginExtension.java [hunk 1]
```diff
@@ -111,12 +111,21 @@ public void setReportsDir(File reportsDir) {
...
         }));
 
         task.getJvmArgumentProviders().add(new JacocoAgent(extension));
+        task.doFirst(new Action<Task>() {
+            @Override
+            public void execute(Task task) {
+                if (extension.isEnabled() && extension.getOutput() == JacocoTaskExtension.Output.FILE) {
+                    File coverageFile = extension.getDestinationFile();
+                    project.delete(coverageFile);
+                }
```
Comment (lptr) [inline subprojects/jacoco/src/main/java/org/gradle/testing/jacoco/plugins/JacocoPluginExtension.java]: Maybe add a comment about why we do it this way?

File: subprojects/jacoco/src/main/java/org/gradle/testing/jacoco/plugins/JacocoPluginExtension.java [hunk 2]
```diff
@@ -126,14 +138,6 @@ public boolean isSatisfiedBy(Task element) {
...
             }
         });
-
-        // Do not cache the Test task if we are appending to the Jacoco output
-        task.getOutputs().doNotCacheIf("JaCoCo agent configured with `append = true`", new Spec<Task>() {
-            @Override
-            public boolean isSatisfiedBy(Task element) {
-                return extension.isEnabled() && extension.isAppend();
-            }
-        });
```
Comment (JLLeitschuh) [inline subprojects/jacoco/src/main/java/org/gradle/testing/jacoco/plugins/JacocoPluginExtension.java]: 🎉

**Comment by TimvdLippe** [comment]
> Note that this is a breaking change: separate tasks now cannot use the
same coverage file, since each of the tasks will delete it.

There does not seem to be any guidance on how to deal with this breaking change. This broke [the Mockito coverage configuration](https://github.com/mockito/mockito/issues/1689) and it is unclear how we can resolve this situation. Could you update the docs or this PR to provide users with a migration strategy?

**Comment by wolfs** [comment]
@TimvdLippe Why can't you use the `JacocoMerge` task and have one execution file per test task?

**Comment by TimvdLippe** [comment]
@wolfs I later figured out that this task exists, but it was no where mentioned in the docs. (I still haven't gotten it to work, even with `JacocoMerge`, but there is probably some way it will eventually work)

**Comment by wolfs** [comment]
@TimvdLippe It seems like you figured it out with the help of @marcphilipp: https://github.com/mockito/mockito/pull/1699

Do you think we should add something to our documentation, so that other users have an easier time migrating?

**Comment by TimvdLippe** [comment]
Yes that would be great. Essentially https://docs.gradle.org/current/userguide/jacoco_plugin.html does not include examples on how to work with the (multiple) `.exec` files. Adding an extra note with the migration example would resolve the confusion for users.

**Pull request merged**